src/local/butler/lint.py

```diff
@@ -15,10 +15,140 @@
 
 import os
 import sys
+import yaml
 
 from local.butler import appengine
 from local.butler import common
 
+_GOLINT_EXCEPTIONS = [
+    'types.go'  # Not all model names conform to Go naming conventions.
+]
+_LICENSE_CHECK_FILENAMES = ['Dockerfile']
+_LICENSE_CHECK_EXTENSIONS = [
+    '.bash',
+    '.c',
+    '.cc',
+    '.cpp',
+    '.css',
+    '.h',
+    '.htm',
+    '.html',
+    '.js',
+    '.go',
+    '.py',
+    '.sh',
+    '.yaml',
+]
+_LICENSE_CHECK_IGNORE_FILENAMES = ['technology.css']
+_LICENSE_CHECK_IGNORE_DIRECTORIES = [
+    'third_party',
+    'templates',  # Generated code.
+]
+_LICENSE_CHECK_STRING = 'http://www.apache.org/licenses/LICENSE-2.0'
+_PY_TEST_SUFFIX = '_test.py'
+_PY_INIT_FILENAME = '__init__.py'
+_YAML_EXCEPTIONS = ['bad.yaml']
+
+
+def license_validate(file_path):
+  """Run license validation for the given source path."""
+  filename = os.path.basename(file_path)
+  extension = os.path.splitext(file_path)[1]
+  if (filename not in _LICENSE_CHECK_FILENAMES and
+      extension not in _LICENSE_CHECK_EXTENSIONS):
+    return
+
+  path_directories = file_path.split(os.sep)
+  for directory in _LICENSE_CHECK_IGNORE_DIRECTORIES:
+    if directory in path_directories:
+      return
+
+  source_filename = os.path.basename(file_path)
+  for check_filename in _LICENSE_CHECK_IGNORE_FILENAMES:
+    if check_filename == source_filename:
+      return
+
+  if _LICENSE_CHECK_STRING in open(file_path).read():
+    return
+
+  print 'Missing license header for %s.' % file_path
+  sys.exit(1)
+
+
+def py_import_order(file_path):
+  """Validate that python imports are alphabetized."""
+
+  def _validate_block(import_block):
+    """Ensure that a single block is ordered properly."""
+    if not import_block:
+      return []
+
+    sorted_import_block = sorted(import_block, key=lambda i: i.lower())
+    if sorted_import_block == import_block:
+      return []
+
+    return ['\n'.join(sorted_import_block)]
+
+  def _validate_imports(data):
+    """Test that a file's contents are ordered properly."""
+    imports = []
+    from_imports = []
+    corrected_import_blocks = []
+    for line in data.splitlines():
+      if line.startswith('import '):
+        imports.append(line)
+      else:
+        corrected_import_blocks += _validate_block(imports)
+        imports = []
+
+      if line.startswith('from '):
```

Comment:
I'm not sure separating `from` blocks and `import` blocks is correct. 
Currently this allows:
```
import b
from a import x
```
Which contradicts the example in the style guide: http://google.github.io/styleguide/pyguide.html#313-imports-formatting.

I think a block should end on a newline.

Is there an open source implementation of this check that we could use instead?

Reply:
can't find any.

Reply:
Should we file a bug then to fix this?

Reply:
filed https://github.com/google/clusterfuzz/issues/340